geomstats/learning/pca.py

```diff
@@ -353,3 +355,186 @@ def _fit(self, X, base_point=None):
+
+    Attributes
+    ----------
+    components_ : array-like, shape=[n_components, 2]
+        Principal axes, representing the directions of maximal variance in the
+        data. They are the initial velocities of the principal geodesics.
+    mean_ : array-like, shape=[2,]
+        Intrinsic mean of the data points.
+
+    References
+    ----------
+    .. [CSV2016] R. Chakraborty, D. Seo, and B. C. Vemuri,
+        "An efficient exact-pga algorithm for constant curvature manifolds."
+        Proceedings of the IEEE conference on computer vision and pattern
+        recognition. 2016.
+    """
+
+    def __init__(self, space, n_vec=100):
+        self.space = space
+        self.half_space = Hyperbolic(2, coords_type="half-space")
+        self.space_ext = Hyperbolic(2, coords_type="extrinsic")
+        self.n_vec = n_vec
+
+    def _project_on_geodesic_extrinsic(self, point_ext, mean_ext, vector_ext):
+        """Project on geodesic in extrinsic coordinates.
+
+        Project point onto geodesic going through mean in direction of vector.
+        """
+        inner_prod_1 = self.space_ext.metric.inner_product(point_ext, vector_ext)
+        inner_prod_2 = self.space_ext.metric.inner_product(point_ext, mean_ext)
+        norm_v = self.space_ext.metric.norm(vector_ext, mean_ext)
+        dist_to_proj = gs.arctanh(-inner_prod_1 / inner_prod_2 / norm_v)
+        proj = gs.einsum("...,i->...i", gs.cosh(dist_to_proj), mean_ext) + gs.einsum(
+            "...,i->...i", gs.sinh(dist_to_proj), vector_ext / norm_v
+        )
+        return proj
+
+    def _fit(self, X, y=None):
+        """Fit the model with X.
+
+        Parameters
+        ----------
+        X : array-like, shape=[..., n_features]
+            Training data in the hyperbolic plane. If the space is
+            the Poincare half-space or Poincare ball, n_features is
+            2. If it is the hyperboloid, n_features is 3.
+        y : Ignored (Compliance with scikit-learn interface)
+
+        Returns
+        -------
+        self : object
+            Returns the instance itself.
+        """
+        estimator = FrechetMean(space=self.space).fit(X)
+        self.mean_ = estimator.estimate_
+
+        mean_half_space = self.space.to_coordinates(self.mean_, "half-space")
+        mean_ext = self.space.to_coordinates(self.mean_, "extrinsic")
+        X_ext = self.space.to_coordinates(X, "extrinsic")
+
+        angles_half_space = gs.linspace(0.0, 2 * gs.pi, self.n_vec)
+        angles_half_space = gs.expand_dims(angles_half_space, axis=1)
+        vectors_half_space = gs.hstack(
+            (gs.cos(angles_half_space), gs.sin(angles_half_space))
+        )
+        norms = self.half_space.metric.norm(vectors_half_space, mean_half_space)
+        vectors_half_space = gs.einsum("ij,i->ij", vectors_half_space, 1 / norms)
+        vectors_ext = self.space.half_space_to_extrinsic_tangent(
+            vectors_half_space, mean_half_space
+        )
```

Comment:
What would be the difference if we get only the last tangent vector from `vectors_half_space`, transform it to extrinsic, get the geodesic function directly from extrinsic, and then sample it uniformly between 0 and 1 with `n_vec`?

Reply:
The difference is that you need to sample from tangent vectors in the extrinsic tangent space and I wanted to go fast so I didn't write down the formula (no need to look for one in the half-space). But it should certainly be changed at some point.